src/cascadia/WindowsTerminal/WindowEmperor.cpp

```diff
@@ -0,0 +1,747 @@
+            findWindowArgs.ResultTargetWindow(targetWindow.WindowId());
+            findWindowArgs.ResultTargetWindowName(targetWindow.WindowName());
+        }
+    });
+
+    _dispatcher = winrt::Windows::System::DispatcherQueue::GetForCurrentThread();
+}
+
+WindowEmperor::~WindowEmperor()
+{
+    _app.Close();
+    _app = nullptr;
+}
+
+void _buildArgsFromCommandline(std::vector<winrt::hstring>& args)
+{
+    if (auto commandline{ GetCommandLineW() })
+    {
+        auto argc = 0;
+
+        // Get the argv, and turn them into a hstring array to pass to the app.
+        wil::unique_any<LPWSTR*, decltype(&::LocalFree), ::LocalFree> argv{ CommandLineToArgvW(commandline, &argc) };
+        if (argv)
+        {
+            for (auto& elem : wil::make_range(argv.get(), argc))
+            {
+                args.emplace_back(elem);
+            }
+        }
+    }
+    if (args.empty())
+    {
+        args.emplace_back(L"wt.exe");
+    }
+}
+
+bool WindowEmperor::HandleCommandlineArgs()
+{
+    std::vector<winrt::hstring> args;
+    _buildArgsFromCommandline(args);
+    auto cwd{ wil::GetCurrentDirectoryW<std::wstring>() };
+
+    Remoting::CommandlineArgs eventArgs{ { args }, { cwd } };
+
+    const auto isolatedMode{ _app.Logic().IsolatedMode() };
+
+    const auto result = _manager.ProposeCommandline(eventArgs, isolatedMode);
+
+    if (result.ShouldCreateWindow())
+    {
+        _createNewWindowThread(Remoting::WindowRequestedArgs{ result, eventArgs });
+
+        _becomeMonarch();
+    }
+    else
+    {
+        const auto res = _app.Logic().GetParseCommandlineMessage(eventArgs.Commandline());
+        if (!res.Message.empty())
+        {
+            AppHost::s_DisplayMessageBox(res);
+            ExitThread(res.ExitCode);
+        }
+    }
+
+    return result.ShouldCreateWindow();
+}
+
+void WindowEmperor::WaitForWindows()
+{
+    MSG message{};
+    while (GetMessageW(&message, nullptr, 0, 0))
+    {
+        TranslateMessage(&message);
+        DispatchMessage(&message);
+    }
+}
+
+void WindowEmperor::_createNewWindowThread(const Remoting::WindowRequestedArgs& args)
+{
+    Remoting::Peasant peasant{ _manager.CreatePeasant(args) };
+
+    auto window{ std::make_shared<WindowThread>(_app.Logic(), args, _manager, peasant) };
+
+    std::weak_ptr<WindowEmperor> weakThis{ weak_from_this() };
+
+    window->Started([weakThis, sender = window]() {
+        // These come in on the sender's thread. Make sure we haven't died
+        // since then
+        if (auto self{ weakThis.lock() })
+        {
+            self->_windowStartedHandler(sender);
+        }
+    });
+
+    window->Exited([weakThis](uint64_t senderID) {
+        // These come in on the sender's thread. Make sure we haven't died
+        // since then
+        if (auto self{ weakThis.lock() })
+        {
+            self->_windowExitedHandler(senderID);
+        }
+    });

```

Comment:
BTW doesn't this leak the `window` instance, because the Started/Exited events are never revoked?

Reply:
I've been coming across this again today and I'd like to propose a solution that might solve this issue categorically. I've previously proposed to move the `std::thread` instantiation into `WindowEmperor.cpp` and I'd like to suggest it again because it would help here.

`WindowThread` basically consists of 2 parts: Things that run before `Started` and things that run after it, but before `Exited`. We can thus split the class into two functions.

This code could be moved into a function called `CreateHost`: https://github.com/microsoft/terminal/blob/9957e5ce52260272c620e0cb6ea5ee998c8177db/src/cascadia/WindowsTerminal/WindowThread.cpp#L22-L28

This code could be moved into a function called `RunMessagePump`: https://github.com/microsoft/terminal/blob/9957e5ce52260272c620e0cb6ea5ee998c8177db/src/cascadia/WindowsTerminal/WindowThread.cpp#L28-L46

Afterwards, `WindowEmperor` can be rewritten like so:
```cpp
void WindowEmperor::_createNewWindowThread(const Remoting::WindowRequestedArgs& args)
{
    Remoting::Peasant peasant{ _manager.CreatePeasant(args) };
    auto window{ std::make_shared<WindowThread>(_app.Logic(), args, _manager, peasant) };
    std::weak_ptr<WindowEmperor> weakThis{ weak_from_this() };

    std::thread t([weakThis, window]() {
        window->CreateHost();

        if (auto self{ weakThis.lock() })
        {
            self->_windowStartedHandler(window);
        }

        window->RunMessagePump();

        if (auto self{ weakThis.lock() })
        {
            self->_windowExitedHandler(senderID);
        }
    });
}
```

The `WindowThread::Start` function is then not necessary anymore and so are the `Start` and `Exited` event handlers. This change effectively "un-inverts" the flow of control and the caller retains full control over what happens. This solves the issue because without an inverted control flow no such references cycles can occur in the first place.

Of course the argument may be made that `WindowThread` now isn't really a thread-owner anymore and this breaks the encapsulation of behavior. But there's no reason it has to be a thread-owner in the first place (since we're actively designing the code). One could thus just rename it to `WindowHostOwner` or something and it would indicate how the "WindowHostOwner" is just another sub-component of `WindowEmperor` itself.

Reply:
_I am a huge fan of this!_

Reply:
BTW I realized that we might have to wrap the contents of the thread lambda in a try/catch, because `std::thread` is specified as:
> The return value of the top-level function is ignored and if it terminates by throwing an exception, [std::terminate](https://en.cppreference.com/w/cpp/error/terminate) is called.

`_windowExitedHandler()` would have to be called inside the try clause as a `wil::scope_exit`. I'm not entirely sure how valuable such a try/catch handler would be though, since exceptions like that should be rare in the first place.

Reply:
I'm gonna file that as a follow-up, just in case. If this does become a hitter during selfhost, we'll get a very loud signal to fix it, but it does seem like something we shouldn't ever encounter....